libdrgn/dwarf_info_cache.c

```diff
@@ -42,6 +42,53 @@ static bool drgn_type_realloc(struct drgn_type **type, size_t capacity,
 	return true;
 }
 
+/*
+ * We use dwarf_formudata() to pull constants out of dwarf attributes.
+ * Unfortunately, GCC's dwarfv2 generator for structure member offsets uses
+ * simple "blocks" instead of the constants that dwarf_formudata() expects.
+ *
+ * So if dwarf_formudata() fails, we handle the special case of the attribute
+ * being a block consisting of a single DW_OP_plus_uconst operation.
+ */
+static int drgn_formudata(Dwarf_Attribute *attr, Dwarf_Word *return_uval)
+{
+	Dwarf_Block block;
+
+	/* dwarf4 common case: constant value */
+	if (dwarf_formudata(attr, return_uval) == 0) {
+		return (0);
+	}
+	/* Try for the dwarf2 case, bailing out if anything looks odd */
+	if (dwarf_formblock(attr, &block) == 0) {
```

Comment:
Can we use `dwarf_getlocation()`? See `drgn_object_from_dwarf_variable()` for an example.

Reply:
We could; my only worry is the overhead of the memory allocation (internal to libdw, which caches them, presumably forever) for each every structure member we ever see.  It looks like it would work, though.  If you're not worried about that overhead, it would certainly make this function simpler.

Reply:
Yeah let's do the simpler way for now. If it ends up causing issues, we can bring back your optimized version.